File: Data Structures & Algorithms/coin-change/submission-3.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Map<Integer, Integer> memo = new HashMap<>();
        int minCoins = dfs(amount, memo, coins);

        if (minCoins >= (int) 1e9) {
            return -1;
        }
        return minCoins;
    }

    private int dfs(int amount, Map<Integer, Integer> memo, int[] coins) {
        if (amount == 0) {
            return 0;
        }
        if (memo.containsKey(amount)) {
            return memo.get(amount);
        }
        int res = (int) 1e9;
        for (int coin: coins) {
            if (amount - coin >= 0) {
                res = Math.min(res, 1 + dfs(amount - coin, memo, coins));
            }
            memo.put(amount, res);
        }
        return res;
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-3.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def dfs(curr: str, left: int, right: int):
            if len(curr) == 2*n:
                res.append(curr)
                return
            
            if left < n:
                dfs(curr + "(", left + 1, right)
            
            if right < left:
                dfs(curr + ")", left, right + 1)
        dfs("", 0, 0)
        return res